python/pyarrow/_compute.pyx

```diff
@@ -904,59 +885,71 @@ class TakeOptions(_TakeOptions):
 
 
 cdef class _PartitionNthOptions(FunctionOptions):
-    def _set_options(self, int64_t pivot):
+    def _set_options(self, pivot):
         self.wrapped.reset(new CPartitionNthOptions(pivot))
 
 
 class PartitionNthOptions(_PartitionNthOptions):
-    def __init__(self, int64_t pivot):
+    def __init__(self, pivot):
         self._set_options(pivot)
 
 
 cdef class _MakeStructOptions(FunctionOptions):
-    def _set_options(self, field_names):
+    def _set_options(self, field_names, field_nullability, field_metadata):
         cdef:
             vector[c_string] c_field_names
-        for n in field_names:
-            c_field_names.push_back(tobytes(n))
-        self.wrapped.reset(new CMakeStructOptions(c_field_names))
+            vector[shared_ptr[const CKeyValueMetadata]] c_field_metadata
+        for name in field_names:
+            c_field_names.push_back(tobytes(name))
+        for metadata in field_metadata:
+            c_field_metadata.push_back(pyarrow_unwrap_metadata(metadata))
+        self.wrapped.reset(
+            new CMakeStructOptions(c_field_names,
+                                   field_nullability,
+                                   c_field_metadata))
 
 
 class MakeStructOptions(_MakeStructOptions):
-    def __init__(self, field_names):
-        self._set_options(field_names)
+    def __init__(self, field_names, field_nullability=None,
```

Comment:
```suggestion
    def __init__(self, field_names, *, field_nullability=None,
```

?

Reply:
`MakeStructOptions` was updated recently with additional mandatory parameters. All arguments are mandatory (see [C++ definition](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/api_scalar.h#L300-L303)), so I think we should leave them positional here as well.

Reply:
Oops, I now noticed that `MakeStructOptions` does have constructor with only the first parameter (`field_names`). I applied the suggested change (keyword-only for optional parameters).